### PR TITLE
Feature 2388 update tci

### LINKED ISSUE #2388: Enhancement: Enhance Terrestrial Coupling Index (TCI) Use Case
*Replace italics below with details for this issue.*

## Describe the Enhancement ##
The TCI use case in the current [METplus User's guide section 7.2.5 (Land Surface)](https://metplus.readthedocs.io/en/latest/generated/model_applications/land_surface/PointStat_fcstCESM_obsFLUXNET2015_TCI.html#sphx-glr-generated-model-applications-land-surface-pointstat-fcstcesm-obsfluxnet2015-tci-py) needs to be enhanced for several aspects. 

First, the current version reads precomputed values of TCI from observations in the Python Embedding script `parm/use_cases/model_applications/land_surface/PointStat_fcstCESM_obsFLUXNET2015_TCI/fluxnet2015_tci.py`  

This python embedding script will need to be upgraded to read the raw surface flux and soil moisture data from FLUXNET2015 data and compute TCI for user specified season (e.g., summer - JJA or winter - DJF).  

Second, the observation data script should have user defined flexibility to compute TCI from both latent and sensible heat flux, as well as multiple soil depths.

Finally, the model data python embedding script should have more flexibility to use different soil layers to match the flexibility of the observation data python script.
`parm/use_cases/model_applications/land_surface/PointStat_fcstCESM_obsFLUXNET2015_TCI/cesm_tci.py`

### Time Estimate ###
2-3 days

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] FLUXNET raw data ingest for the observation data python embedding script
- [ ] Modifications to observation data python embedding script and METplus configuration for user selected variables
- [ ] Modifications to model data python embedding script for user selected variables 
- [ ] Add a description of this data into the METplus Verification Datasets Guide.

### Relevant Deadlines ###
12/31/2023

### Funding Source ###
7790111 (CLASP)

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED CYCLE ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by DanielAdriaansen**
Closes #2388. 

Includes:
1. New table of contents at top of use case docs file
2. Updated FCST python embedding script
3. Updated OBS Python embedding script
4. New station metadata file required for the OBS data
5. Reorganized and updated use case conf file
6. New OBS sample data
 
- [ ] Describe testing already performed for these changes:</br>
I ran the use case locally and it succeeds, and it also succeds in GHA.

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
1. Try running the use case locally
2. @anewman89, since I am not using pre-computed TCI and using raw FLUXNET, we should probably verify that the TCI values are correct somehow. However we can't use the old TCI values because we can't guarantee that the same data filtering was applied that we are applying to the raw data. I suppose that since FCST and OBS use the same function, and the FCST input data didn't change, we could take a difference between the new FCST TCI grid and the old FCST TCI grid values- then we'd at least know the function produces the correct values.

- [ ] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]** 
Yes

- [ ] Do these changes include sufficient testing updates? **[Yes or No]** 
Yes

- [ ] Will this PR result in changes to the test suite? **[Yes or No]**</br> If **yes**, describe the new output and/or changes to the existing output:</br> 
Yes, the OBS input data are different and therefore the OBS values being used by MET are also different, which will result in differences in the test data. After the PR is merged, the test data will need to be updated.

- [ ] Do these changes introduce new SonarQube findings? **[Yes or No]**</br>
If **yes**, please describe:
Yes, but I resolved 3 code smells in the Python embedding script before SonarQube licensing issues cropped up, so ultimately there should be no new SonarQube findings.

- [ ] Please complete this pull request review by **[Fill in date]**.</br>
At your earliest convenience.

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by georgemccabe** [comment]
@DanielAdriaansen, could you fill out the pull request template? It would be useful to know what testing is expected of the reviewers and if the differences in the use case output are expected.

**Comment by DanielAdriaansen** [comment]
> @DanielAdriaansen, could you fill out the pull request template? It would be useful to know what testing is expected of the reviewers and if the differences in the use case output are expected.

Done, sorry about that. I had this in draft and converted it as ready but forgot to go back and add the details.

**Comment by georgemccabe** [comment]
[Documentation changes](https://metplus--2550.org.readthedocs.build/en/2550/generated/model_applications/land_surface/PointStat_fcstCESM_obsFLUXNET2015_TCI.html#sphx-glr-generated-model-applications-land-surface-pointstat-fcstcesm-obsfluxnet2015-tci-py) from this PR

File: docs/use_cases/model_applications/land_surface/PointStat_fcstCESM_obsFLUXNET2015_TCI.py [hunk 1]
```diff
@@ -5,50 +5,84 @@
...
-# | **Forecast:** CESM 1979-1983 Simulations a. Community Land Model (CLM) and b. Community Atmosphere Model (CAM) 
-#
-# | **Observations:** FLUXNET2015 post processed and converted to NETCDF. This data includes data collected from multiple regional flux towers.
+# | **Forecast:** CESM 1979-1983 Simulations 
+# | * Community Land Model (CLM) file
+# | * Community Atmosphere Model (CAM) file
 #
+# | **Observations:** Raw FLUXNET2015 observations
 #
 # | **Location:** All of the input data required for this use case can be found in the met_test sample data tarball. Click here to the METplus releases page and download sample data for the appropriate release: https://github.com/dtcenter/METplus/releases
```
Comment (georgemccabe) [inline docs/use_cases/model_applications/land_surface/PointStat_fcstCESM_obsFLUXNET2015_TCI.py]:
```suggestion
# | **Location:** All of the input data required for this use case can be found in the land_surface sample data tarball. Click here to the METplus releases page and download sample data for the appropriate release: https://github.com/dtcenter/METplus/releases
```

**Comment by DanielAdriaansen** [comment]
Looking closer at the difference tests from CI, I note that there were unexpected differences in the FCST data. However, way back I fixed a bug here: https://github.com/dtcenter/METplus/issues/2388#issuecomment-1787789180, and the differences were only in MAM and JJA so that is reassuring and DJF and SON show zero differences. Thus, the formulation of TCI remains the same, and is being called correctly for the FCST.

I spot checked a few OBS sites. From the `point_stat` MPR files for DJF, I see:
```
Truth:
Borgo Cioffi         40.52375   14.95744      10       0   4.62984    0.42177
Tonzi Ranch        38.4316  -120.96598      10       0  -2.23139    1.7888
Grignon               48.84422    1.95191      10       0  -1.09105    1.2232

Output:
IT-BCi    40.5237   14.9574       10       0   4.62984    0.95305
US-Ton    38.4309 -120.966        10       0  -2.23139    2.91969
FR-Gri    48.8442    1.9519       10       0  -1.09105    0.717
```
What we can see is that:
1) FCST values match
2) OBS values do not match, but the sign is the same and the new values are higher for these three sites
3) The station metadata (lat/lon) matches

The OBS values being different here is almost certainly driven by differences in decisions about what quality of raw FLUXNET data to include in the TCI calculation, compared to whatever data the previous TCI provider chose to use. I'm not sure there's a way to quantify this, but rather just be confident in our data filtering and take the matching FCST values as an indicator the actual math for computing TCI remains unchanged.

**Review by georgemccabe — CHANGES_REQUESTED** [review]

**Review by georgemccabe — APPROVED** [review]
I added the new data for the v6.0 and develop branches, then confirmed that I could run the use case on seneca using the new data.

I approve, pending @anewman89's review

**Comment by anewman89** [comment]
@DanielAdriaansen Thanks for the testing and explanation of the results.  I agree it seems like we're ok with the TCI computation. I also approve the PR.

**Review by anewman89 — APPROVED** [review]

**Pull request merged**